src/contexts/ethereum-provider.context.tsx

```diff
@@ -0,0 +1,120 @@
+import { createContext, FC, useContext, useEffect, useState } from "react";
+import { Web3Provider } from "@ethersproject/providers";
+import WalletConnectProvider from "@walletconnect/ethereum-provider";
+import { useNavigate } from "react-router-dom";
+
+import { EthereumEvent, WalletName } from "src/domain";
+import { AsyncTask } from "src/utils/types";
+import { ethereumAccountsParser } from "src/adapters/parsers";
+import { getConnectedAccounts } from "src/adapters/ethereum";
+import { useEnvContext } from "./env.context";
+import routes from "src/routes";
+
+interface EthereumProviderContextData {
+  provider?: Web3Provider;
+  account: AsyncTask<string, string>;
+  connectProvider?: (walletName: WalletName) => Promise<void>;
+}
+
+const EthereumProviderContext = createContext<EthereumProviderContextData>({
+  provider: undefined,
+  account: { status: "pending" },
+});
+
+const EthereumProviderProvider: FC = (props) => {
+  const navigate = useNavigate();
+  const [provider, setProvider] = useState<Web3Provider>();
+  const [account, setAccount] = useState<AsyncTask<string, string>>({ status: "pending" });
+  const env = useEnvContext();
+
+  const connectProvider = async (walletName: WalletName): Promise<void> => {
+    switch (walletName) {
+      case WalletName.METAMASK: {
+        if (window.ethereum) {
+          const web3Provider = new Web3Provider(window.ethereum);
+
+          return getConnectedAccounts(web3Provider)
+            .then((accounts) => {
+              setProvider(web3Provider);
+              setAccount({ status: "successful", data: accounts[0] });
+            })
+            .catch(() => setAccount({ status: "failed", error: "Error loading account" }));
+        } else {
+          return setAccount({
+            status: "failed",
+            error: "There is no Ethereum provider injected in the window",
```

Comment:
Wouldn't it make sense to have a meaningful error that we can show to the user? Something in the line of:

> We cannot detect your wallet. Make sure the MetaMask extension is installed and active in your browser.

Reply:
Yes, this is going to be done in the first PR. I still need to figure out how are we going to manage the `chainId` as we need to take into account also if we are on mainnet environment or testnet environment.